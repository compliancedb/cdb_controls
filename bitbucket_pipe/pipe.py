import sys
from bitbucket_pipes_toolkit import Pipe, get_logger

import cdb.control_junit
import cdb.create_release
import cdb.put_artifact
import cdb.put_artifact_image
import cdb.put_pipeline
from cdb.bitbucket import get_bitbucket_repo_url, adapt_put_artifact_image_env_variables, \
    adapt_create_release_variables, adapt_control_junit_env_variables, adapt_put_artifact_env_variables

sys.path.append('/app/')
import cdb.cdb_utils
import cdb.bitbucket
from cdb.cdb_utils import set_artifact_sha_env_variable_from_file_or_image


schema = {
    'CDB_COMMAND': {'type': 'string', 'required': True},
    'CDB_PIPELINE_DEFINITION': {'type': 'string', 'required': True},
    'CDB_API_TOKEN': {'type': 'string', 'required': True},
    # PUT Pipeline Variables
    #   ...none...
    # Put Artifact Variables
    'CDB_IS_COMPLIANT': {'type': 'boolean', 'required': False},
    'CDB_ARTIFACT_GIT_URL': {'type': 'string', 'required': False},
    'CDB_ARTIFACT_GIT_COMMIT': {'type': 'string', 'required': False},
    'CDB_CI_BUILD_URL': {'type': 'string', 'required': False},
    'CDB_BUILD_NUMBER': {'type': 'string', 'required': False},
    'CDB_ARTIFACT_SHA': {'type': 'string', 'required': False},
    'CDB_ARTIFACT_FILENAME': {'type': 'string', 'required': False},

}

logger = get_logger()


class BitbucketPipe(Pipe):
    def run(self):
        super().run()

        logger.info('Executing the pipe...')

        command = self.get_variable('CDB_COMMAND')
        pipeline_definition_file = self.get_variable('CDB_PIPELINE_DEFINITION')
        api_token = self.get_variable('CDB_API_TOKEN')

        print("Command: " + command)
        print("Pipeline: " + pipeline_definition_file)
        print("API_TOKEN: " + api_token)

        if command == "put_pipeline":
            cdb.put_pipeline.put_pipeline(pipeline_definition_file)
        if command == "put_artifact":
            adapt_put_artifact_env_variables()
            cdb.put_artifact.put_artifact(pipeline_definition_file)
        if command == "put_artifact_image":
            adapt_put_artifact_image_env_variables()
            cdb.put_artifact_image.put_artifact_image(pipeline_definition_file)
        if command == "control_junit":
            adapt_control_junit_env_variables()
            cdb.control_junit.control_junit(pipeline_definition_file)
        if command == "control_bitbucket_pr":
            set_artifact_sha_env_variable_from_file_or_image()
            cdb.bitbucket.put_bitbucket_pull_request(pipeline_definition_file)
        if command == "create_release":
            adapt_create_release_variables()
            cdb.create_release.create_release(pipeline_definition_file)

        self.success(message="Success!")


if __name__ == '__main__':
    pipe = BitbucketPipe(pipe_metadata='/pipe.yml', schema=schema)
    pipe.run()
