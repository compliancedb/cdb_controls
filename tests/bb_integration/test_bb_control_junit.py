from pipe import BitbucketPipe, schema

from tests.utils import *


def test_required_env_vars(capsys, mocker):
    # BITBUCKET_COMMIT and BITBUCKET_BUILD_NUMBER are required
    # If they are not set the code tries to set an env-var to none
    # BITBUCKET_WORKSPACE probably should be required, but None ends up in the url
    # BITBUCKET_REPO_SLUG probably should be required, but None ends up in the url
    # CDB_TEST_RESULTS_DIR has a default, '/data/junit/' but this is _not_ checked
    # resulting in message:
    #   "JUnit results xml verified by compliancedb/cdb_controls: All tests passed in 0 test suites"
    bitbucket_commit = "abc50c8a53f79974d615df335669b59fb56a4ed3"
    env = {
        "CDB_COMMAND": "control_junit",
        "CDB_PIPELINE_DEFINITION": "tests/data/pipefile.json",
        "CDB_API_TOKEN": "SOME_RANDOM_TOKEN",
        "CDB_ARTIFACT_SHA": "aacdaef69c676c2466571d3211110d559ccc2032b258fc5e73f99a103db462ee",
        "BITBUCKET_COMMIT": bitbucket_commit,
        "BITBUCKET_BUILD_NUMBER": "127",
    }
    set_env_vars = {
        'CDB_ARTIFACT_GIT_URL': 'https://bitbucket.org/None/None/commits/' + bitbucket_commit,
        'CDB_ARTIFACT_GIT_COMMIT': bitbucket_commit,
        'CDB_BUILD_NUMBER': '127',
        'CDB_CI_BUILD_URL': 'https://bitbucket.org/None/None/addon/pipelines/home#!/results/127'
    }
    with dry_run(env, set_env_vars):
        pipe = BitbucketPipe(pipe_metadata='/pipe.yml', schema=schema)
        pipe.run()
    verify_approval(capsys)


def test_all_env_vars(capsys, mocker):
    bitbucket_commit = "dad50c8a53f79974d615df335669b59fb56a4ed3"
    env = {
        "CDB_COMMAND": "control_junit",
        "CDB_PIPELINE_DEFINITION": "tests/data/pipefile.json",
        "CDB_API_TOKEN": "SOME_RANDOM_TOKEN",
        "CDB_ARTIFACT_SHA": "aacdaef69c676c2466571d3277770d559ccc2032b258fc5e73f99a103db462ee",
        "CDB_TEST_RESULTS_DIR": "/app/tests/data/control_junit/xml-with-fails",
        "BITBUCKET_COMMIT": bitbucket_commit,
        "BITBUCKET_BUILD_NUMBER": "703",
        "BITBUCKET_WORKSPACE": "acme",
        "BITBUCKET_REPO_SLUG": "beep-beep"
    }
    set_env_vars = {
        'CDB_ARTIFACT_GIT_URL': 'https://bitbucket.org/acme/beep-beep/commits/dad50c8a53f79974d615df335669b59fb56a4ed3',
        'CDB_ARTIFACT_GIT_COMMIT': 'dad50c8a53f79974d615df335669b59fb56a4ed3',
        'CDB_BUILD_NUMBER': '703',
        'CDB_CI_BUILD_URL': 'https://bitbucket.org/acme/beep-beep/addon/pipelines/home#!/results/703'
     }
    with dry_run(env, set_env_vars):
        pipe = BitbucketPipe(pipe_metadata='/pipe.yml', schema=schema)
        pipe.run()
    verify_approval(capsys)


