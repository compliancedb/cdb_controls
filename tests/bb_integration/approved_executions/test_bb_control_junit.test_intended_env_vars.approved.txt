Command: control_junit
Pipeline: tests/data/pipefile.json
API_TOKEN: SOME_RANDOM_TOKEN
Publish evidence to ComplianceDB
{'evidence_type': 'junit', 'contents': {'is_compliant': True, 'url': 'https://bitbucket.org/acme/road-runner/addon/pipelines/home#!/results/127', 'description': 'JUnit results xml verified by compliancedb/cdb_controls: All tests passed in 0 test suites'}}
Docker image digest found from environment variable
IMAGE DIGEST: aacdaef69c676c2466571d3211110d559ccc2032b258fc5e73f99a103db462ee
Putting this payload:
{
    "contents": {
        "description": "JUnit results xml verified by compliancedb/cdb_controls: All tests passed in 0 test suites",
        "is_compliant": true,
        "url": "https://bitbucket.org/acme/road-runner/addon/pipelines/home#!/results/127"
    },
    "evidence_type": "junit"
}
To this url: https://app.compliancedb.com/api/v1/projects/merkely/test-pipefile/artifacts/aacdaef69c676c2466571d3211110d559ccc2032b258fc5e73f99a103db462ee
DRY RUN: Put not sent
[32mâœ” Success![0m
